pyccl/nl_pt/ept.py

```diff
@@ -0,0 +1,700 @@
+            self.one_loop_dd = self.dd_bias[0:1]
+            self.with_matter_1loop = True
+        elif self.with_matter_1loop:  # Only 1-loop matter needed
+            self.one_loop_dd = self.pt.one_loop_dd(
+                pklz0, P_window=self.fastpt_par['P_window'],
+                C_window=self.fastpt_par['C_window'])
+
+        # Intrinsic alignment templates
+        if self.with_IA:
+            self.ia_ta = self.pt.IA_ta(
+                pklz0, P_window=self.fastpt_par['P_window'],
+                C_window=self.fastpt_par['C_window'])
+            self.ia_tt = self.pt.IA_tt(
+                pklz0, P_window=self.fastpt_par['P_window'],
+                C_window=self.fastpt_par['C_window'])
+            self.ia_mix = self.pt.IA_mix(
+                pklz0, P_window=self.fastpt_par['P_window'],
+                C_window=self.fastpt_par['C_window'])
+
+        # b1/bk power spectrum
+        pks = {}
+        if 'nonlinear' in [self.b1_pk_kind, self.bk2_pk_kind]:
+            pks['nonlinear'] = np.array([cosmo.nonlin_matter_power(self.k_s, a)
+                                         for a in self.a_s])
+        if 'linear' in [self.b1_pk_kind, self.bk2_pk_kind]:
+            pks['linear'] = np.array([cosmo.linear_matter_power(self.k_s, a)
+                                      for a in self.a_s])
+        if 'pt' in [self.b1_pk_kind, self.bk2_pk_kind]:
+            if 'linear' in pks:
+                pk = pks['linear']
+            else:
+                pk = np.array([cosmo.linear_matter_power(self.k_s, a)
+                               for a in self.a_s])
+            # Add SPT correction
+            pk += self._g4[:, None] * self.one_loop_dd[0][None, :]
+            pks['pt'] = pk
+        self.pk_b1 = pks[self.b1_pk_kind]
+        self.pk_bk = pks[self.bk2_pk_kind]
+
+        # Reset template power spectra
+        self._pk2d_temp = {}
+        self._pt_init = True
+
+    def _get_pgg(self, tr1, tr2):
+        """ Get the number counts auto-spectrum at the internal
+        set of wavenumbers and scale factors.
+
+        Args:
+            tr1 (:class:`~pyccl.nl_pt.tracers.PTTracer`): first
+                tracer to correlate.
+            tr2 (:class:`~pyccl.nl_pt.tracers.PTTracer`): first
+                tracer to correlate.
+
+        Returns:
+            array_like: 2D array of shape `(N_a, N_k)`, where `N_k` \
+                is the size of this object's `k_s` attribute, and \
+                `N_a` is the size of the object's `a_s` attribute.
+        """
+        self._check_pt_init()
+        # Get Pk templates
+        Pd1d1 = self.pk_b1
+        Pd1d2 = self._g4[:, None] * self.dd_bias[2][None, :]
+        Pd2d2 = self._g4[:, None] * self.dd_bias[3][None, :]
+        Pd1s2 = self._g4[:, None] * self.dd_bias[4][None, :]
+        Pd2s2 = self._g4[:, None] * self.dd_bias[5][None, :]
+        Ps2s2 = self._g4[:, None] * self.dd_bias[6][None, :]
+        Pd1p3 = self._g4[:, None] * self.dd_bias[8][None, :]
```

Comment:
This creates a new axis for `self._g4` every time. Factor it out and do it only once.

Reply:
Done